File: plugins/cordova-plugin-muzik/src/android/ChargeStatusBridge.java
package com.muzik.androidplugin;

import com.muzik.accessory.MzAccessory;
import com.muzik.accessory.callback.IChargeStatusCallback;

import java.util.LinkedList;
import java.util.Queue;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.PluginResult;

public class ChargeStatusBridge implements IChargeStatusCallback {

    private MzAccessory mzAccessory;
    private Queue<CallbackContext> callbacks;

    public ChargeStatusBridge(MzAccessory mzAccessory) {
        this.mzAccessory = mzAccessory;
        this.callbacks = new LinkedList<CallbackContext>();
    }

    public synchronized void getChargeStatus(CallbackContext callback) {
        this.callbacks.add(callback);
        this.mzAccessory.getChargeStatus(this);
    }
    
    @Override
    public synchronized void onResponseReceived(boolean charging) {
        if (!this.callbacks.isEmpty()) {
            PluginResult result = new PluginResult(PluginResult.Status.OK, charging);
            this.callbacks.poll().sendPluginResult(result);
        }
    }
}

File: cordova-plugin-muzik/src/ios/src/ChargeStatusBridge.m
//
//  ChargeStatusBridge.h
//

#import "ChargeStatusBridge.h"
#import "Queue.h"

@interface ChargeStatusBridge ()

@property (nonatomic, strong) MZAccessory* mzAccessory;
@property (nonatomic, strong) CDVPlugin* plugin;
@property (nonatomic, strong) Queue* commandQueue;

@end

@implementation ChargeStatusBridge

// Default constructor
- (id) init {
    self = [super init];
    if (self) {
        self.mzAccessory = nil;
        self.plugin = nil;
        self.commandQueue = [[Queue alloc] init];
    }
    return self;
}

// Constructor with references to an instance of MZAccessory, and CDVPlugin
- (id) initWithAccessory:(MZAccessory *)mzAccessory plugin:(CDVPlugin*)plugin{
    self = [super init];
    if (self) {
        self.mzAccessory = mzAccessory;
        self.plugin = plugin;
        self.commandQueue = [[Queue alloc] init];
    }
    return self;
}

// Get charge status
- (void) getChargeStatus:(CDVInvokedUrlCommand *)command {
    if (command == nil){
        return;
    }
    @synchronized
    (self) {
        [self.commandQueue add:command];
    }
    MZChargeStatusCallback callback = ^void(BOOL charging){
        CDVInvokedUrlCommand* callbackCommand = nil;
        @synchronized
        (self) {
            if (![self.commandQueue isEmpty]) {
                callbackCommand = [self.commandQueue poll];
            }
        }
        if (callbackCommand != nil){
            CDVPluginResult* pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsBool:charging];
            [self.plugin.commandDelegate sendPluginResult:pluginResult callbackId:callbackCommand.callbackId];
        }
    };
    [self.mzAccessory getChargeStatusWithBlock:callback];
}

@end